File: Delphi/Ginkgo SPI/Delphi_USB_SPI_Test/ControlSPI.pas
unit ControlSPI;

interface
const
//Device type definition
VAI_USBADC	=	1;
VCI_USBCAN1	=	3;
VCI_USBCAN2	=	4;
VGI_USBGPIO	=	1;
VII_USBI2C	=	1;
VSI_USBSPI	=	2;
//错误码定义
ERR_SUCCESS					      =	(0);		// 没有错误
ERR_PARAMETER_NULL			  =	(-1);	  // 传入的指针为空指针
ERR_INPUT_DATA_TOO_MUCH		=	(-2);	  // 参数输入个数多余规定个数
ERR_INPUT_DATA_TOO_LESS		=	(-3);	  // 参数输入个数少余规定个数
ERR_INPUT_DATA_ILLEGALITY	=	(-4);	  // 参数传入格式和规定的不符合
ERR_USB_WRITE_DATA			  =	(-5);	  // USB写数据错误
ERR_USB_READ_DATA			    =	(-6);	  // USB读数据错误
ERR_READ_NO_DATA			    =	(-7);	  // 请求读数据时返回没有数据
ERR_OPEN_DEVICE				    =	(-8);	  // 打开设备失败
ERR_CLOSE_DEVICE			    =	(-9);	  // 关闭设备失败
ERR_EXECUTE_CMD				    =	(-10);	// 设备执行命令失败
ERR_SELECT_DEVICE			    =	(-11);	// 选择设备失败
ERR_DEVICE_OPENED			    =	(-12);	// 设备已经打开
ERR_DEVICE_NOTOPEN			  =	(-13);	// 设备没有打开
ERR_BUFFER_OVERFLOW			  =	(-14);	// 缓冲区溢出
ERR_DEVICE_NOTEXIST			  =	(-15);	// 此设备不存在
ERR_LOAD_KERNELDLL			  =	(-16);	// 装载动态库失败
ERR_CMD_FAILED				    =	(-17);	// 执行命令失败错误码
ERR_BUFFER_CREATE			    =	(-18);	// 内存不足
//Ginkgo系列适配器信息的数据类型。
type
PVSI_BOARD_INFO = ^VSI_BOARD_INFO;
VSI_BOARD_INFO = record
	ProductName:Array[0..31] Of Byte;	    //硬件名称，比如“Ginkgo-SPI-Adaptor”（注意：包括字符串结束符‘\0’）
	FirmwareVersion:Array[0..3] Of Byte;	//固件版本
	HardwareVersion:Array[0..3] Of Byte;	//硬件版本
	SerialNumber:Array[0..11] Of Byte;	  //适配器序列号
end;
//USB-SPI初始化结构体
type
PVSI_INIT_CONFIG = ^VSI_INIT_CONFIG;
VSI_INIT_CONFIG = record
  ControlMode:Byte;     //SPI控制方式:0-硬件控制（全双工模式）,1-硬件控制（半双工模式），2-软件控制（半双工模式）,3-单总线模式，数据线输入输出都为MOSI
  TranBits:Byte;        //数据传输字节宽度，在8和16之间取值
  MasterMode:Byte;      //主从选择控制:0-从机，1-主机
  CPOL:Byte;            //时钟极性控制:0-SCK高有效，1-SCK低有效
  CPHA:Byte;            //时钟相位控制:0-第一个SCK时钟采样，1-第二个SCK时钟采样
  LSBFirst:Byte;        //数据移位方式:0-MSB在前，1-LSB在前
  SelPolarity:Byte;     //片选信号极性:0-低电平选中，1-高电平选中
  ClockSpeed:Integer;   //SPI时钟频率:单位为HZ
end;
//函数声明
function VSI_ScanDevice(NeedInit:Byte):Integer; stdcall;
function VSI_OpenDevice(DevType,DevIndex,Reserved:Integer):Integer; stdcall;
function VSI_CloseDevice(DevType,DevIndex:Integer):Integer; stdcall;
function VSI_ReadBoardInfo(DevIndex:Integer;pInfo:PVSI_BOARD_INFO):Integer; stdcall;
function VSI_InitSPI(DevType, DevIndex:Integer; pInitConfig:PVSI_INIT_CONFIG):Integer; stdcall;
function VSI_WriteBytes(DevType, DevIndex, SPIIndex:Integer;pWriteData:PByte;Len:Word):Integer; stdcall;
function VSI_ReadBytes(DevType, DevIndex, SPIIndex:Integer;pReadData:PByte;Len:Word):Integer; stdcall;
function VSI_WriteReadBytes(DevType, DevIndex, SPIIndex:Integer;pWriteData:PByte;WriteLen:Word;pReadData:PByte;ReadLen:Word):Integer; stdcall;
function VSI_WriteBits(DevType, DevIndex, SPIIndex:Integer;pWriteBitStr:PChar):Integer; stdcall;
function VSI_ReadBits(DevType, DevIndex, SPIIndex:Integer;pReadBitStr:PChar;ReadBitsNum:Integer):Integer; stdcall;
function VSI_WriteReadBits(DevType, DevIndex, SPIIndex:Integer;pWriteBitStr,pReadBitStr:PChar;ReadBitsNum:Integer):Integer; stdcall;
function VSI_SlaveReadBytes(DevType,DevIndex:Integer;pReadData:PByte;pBytesNum:PInteger;WaitTime:Integer):Integer; stdcall;
function VSI_SlaveWriteBytes(DevType,DevIndex:Integer;pWriteData:PByte;WriteBytesNum:Integer):Integer; stdcall;
function VSI_SetUserKey(DevType,DevIndex:Integer;pUserKey:PByte):Integer; stdcall;
function VSI_CheckUserKey(DevType,DevIndex:Integer;pUserKey:PByte):Integer; stdcall;

implementation

function VSI_ScanDevice;external 'Ginkgo_Driver.dll' name 'VSI_ScanDevice';
function VSI_OpenDevice;external 'Ginkgo_Driver.dll' name 'VSI_OpenDevice';
function VSI_CloseDevice;external 'Ginkgo_Driver.dll' name 'VSI_CloseDevice';
function VSI_ReadBoardInfo;external 'Ginkgo_Driver.dll' name 'VSI_ReadBoardInfo';
function VSI_InitSPI;external 'Ginkgo_Driver.dll' name 'VSI_InitSPI';
function VSI_WriteBytes;external 'Ginkgo_Driver.dll' name 'VSI_WriteBytes';
function VSI_ReadBytes;external 'Ginkgo_Driver.dll' name 'VSI_ReadBytes';
function VSI_WriteReadBytes;external 'Ginkgo_Driver.dll' name 'VSI_WriteReadBytes';
function VSI_WriteBits;external 'Ginkgo_Driver.dll' name 'VSI_WriteBits';
function VSI_ReadBits;external 'Ginkgo_Driver.dll' name 'VSI_ReadBits';
function VSI_WriteReadBits;external 'Ginkgo_Driver.dll' name 'VSI_WriteReadBits';
function VSI_SlaveReadBytes;external 'Ginkgo_Driver.dll' name 'VSI_SlaveReadBytes';
function VSI_SlaveWriteBytes;external 'Ginkgo_Driver.dll' name 'VSI_SlaveWriteBytes';
function VSI_SetUserKey;external 'Ginkgo_Driver.dll' name 'VSI_SlaveWriteBytes';
function VSI_CheckUserKey;external 'Ginkgo_Driver.dll' name 'VSI_CheckUserKey';
end.

File: Visual Basic.Net/Ginkgo SPI/VB.NET_USB_SPI_M95040/USB_SPI_M95040/USB_SPI_M95040/ControlSPI.vb
Option Strict Off
Option Explicit On

Imports System.Runtime.InteropServices
Module ControlSPI
    'General Error Code
    Public Const SUCCESS As Int32 = 0                   ' no Error
    Public Const PARAMETER_NULL As Int32 = -1           ' void pointer 
    Public Const INPUT_DATA_TOO_MUCH As Int32 = -2      ' too many parameters
    Public Const INPUT_DATA_TOO_LESS As Int32 = -3      ' too few parameters
    Public Const INPUT_DATA_ILLEGALITY As Int32 = -4    ' illegal parameter
    Public Const USB_WRITE_DATA_ERROR As Int32 = -5     ' USB write data error
    Public Const USB_READ_DATA_ERROR As Int32 = -6      ' USB read data error
    Public Const READ_NO_DATA As Int32 = -7             ' no data return when request data
    Public Const OPEN_DEVICE_FAILD As Int32 = -8        ' failed to open device
    Public Const CLOSE_DEVICE_FAILD As Int32 = -9       ' failed to close device
    Public Const EXECUTE_CMD_FAILD As Int32 = -10       ' the command failed to execute  
    Public Const SELECT_DEVICE_FAILD As Int32 = -11     ' failed to select device
    Public Const DEVICE_OPENED As Int32 = -12           ' device has open
    Public Const DEVICE_NOTOPEN As Int32 = -13          ' device not open
    Public Const BUFFER_OVERFLOW As Int32 = -14         ' buffer overflow
    Public Const DEVICE_NOTEXIST As Int32 = -15         ' device not exist
    Public Const LOAD_KERNELDLL As Int32 = -16          ' failed to load KernelDLL
    Public Const CMD_FAILED As Int32 = -17              ' failed to execute command
    Public Const BUFFER_CREATE As Int32 = -18           ' out of memory
    ' Define Adapter type
    Public Const VSI_USBSPI As Int32 = 2
    ' The adapter type definition and data initialization definition
    Public Structure VSI_INIT_CONFIG
        Dim ControlMode As Byte     ' SPI control mode: 0->hardware control(full duplex) 1->hardware control(half duplex) 2->software control(half duplex) 3-> one wire mode
        Dim TranBits As Byte        ' Width  of data (between 8 and 16)
        Dim MasterMode As Byte      ' Master mode : 0 -> slave mode 1 -> master mode
        Dim CPOL As Byte            ' Clock Polarity: 0 -> SCK active-high  1->SCK active-low 
        Dim CPHA As Byte            ' Clock Phase: 0 -> sample on the leading (first) clock edge 1-> sample on the trailing (second) clock edge
        Dim LSBFirst As Byte        ' whether or not LSB first: 0->MSB first 1-> LSB first
        Dim SelPolarity As Byte     ' Chip select Polarity: 0-> low to select 1-> high to select
        Dim ClockSpeed As UInt32    ' SPI clock frequency(in Hz), 36000000,18000000,9000000,4500000,2250000,1125000,562500,281250 available in hardware mode，1KHz to 1MHz available in software mode
    End Structure
    ' Adapter information data type
    Public Structure VSI_BOARD_INFO
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=32)> _
        Dim ProductName As Byte()        'Product Name
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=4)> _
        Dim FirmwareVersion As Byte()    'Firmware Version
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=4)> _
        Dim HardwareVersion As Byte()    'Hardware Version
        <MarshalAs(UnmanagedType.ByValArray, SizeConst:=12)> _
        Dim SerialNumber As Byte()       'Serial Number
    End Structure
    ' Function
    Declare Function VSI_ScanDevice Lib "Ginkgo_Driver.dll" (ByVal NeedInit As Byte) As Int32
    Declare Function VSI_OpenDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal Reserved As Int32) As Int32
    Declare Function VSI_CloseDevice Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32) As Int32
    Declare Function VSI_ReadBoardInfo Lib "Ginkgo_Driver.dll" (ByVal DevIndex As Int32, ByRef pInfo As VSI_BOARD_INFO) As Int32
    Declare Function VSI_InitSPI Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByRef pInitConfig As VSI_INIT_CONFIG) As Int32
    Declare Function VSI_WriteBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteData() As Byte, ByVal WriteLen As UInt16) As Int32
    Declare Function VSI_ReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pReadData() As Byte, ByVal ReadLen As UInt16) As Int32
    Declare Function VSI_WriteReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteData() As Byte, ByVal WriteLen As UInt16, ByVal pReadData() As Byte, ByVal ReadLen As UInt16) As Int32
    ' Only available in software mode 
    Declare Function VSI_WriteBits Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteBitStr() As Byte) As Int32
    Declare Function VSI_ReadBits Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pReadBitStr() As Byte, ByVal ReadBitsNum As UInt32) As Int32
    Declare Function VSI_WriteReadBits Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteBitStr() As Byte, ByVal pReadBitStr() As Byte, ByVal ReadBitsNum As UInt32) As Int32
    ' Slave mode function
    Declare Function VSI_SlaveReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pReadData() As Byte, ByRef pBytesNum As Int32, ByVal WaitTime As Int32) As Int32
    Declare Function VSI_SlaveWriteBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pWriteData() As Byte, ByVal WriteBytesNum As Int32) As Int32
    ' Cryptographic function
    Declare Function VSI_SetUserKey Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pUserKey() As Byte) As Int32
    Declare Function VSI_CheckUserKey Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal pUserKey() As Byte) As Int32
    ' Block option function
    Declare Function VSI_BlockWriteBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteData() As Byte, ByVal BlockSize As UInt16, ByVal BlockNum As UInt16, ByVal IntervalTime As UInt32) As Int32
    Declare Function VSI_BlockReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pReadData() As Byte, ByVal BlockSize As UInt16, ByVal BlockNum As UInt16, ByVal IntervalTime As UInt32) As Int32
    Declare Function VSI_BlockWriteReadBytes Lib "Ginkgo_Driver.dll" (ByVal DevType As Int32, ByVal DevIndex As Int32, ByVal SPIIndex As Int32, ByVal pWriteData() As Byte, ByVal WriteBlockSize As UInt16, ByVal pReadData() As Byte, ByVal ReadBlockSize As UInt16, ByVal BlockNum As UInt16, ByVal IntervalTime As UInt32) As Int32

End Module
